tests/cpgrid/cuboidShape_test.cpp

```diff
@@ -0,0 +1,122 @@
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "config.h"
+
+#define BOOST_TEST_MODULE CuboidShapeTest
+#include <boost/test/unit_test.hpp>
+#include <boost/version.hpp>
+#if BOOST_VERSION / 100000 == 1 && BOOST_VERSION / 100 % 1000 < 71
+#include <boost/test/floating_point_comparison.hpp>
+#else
+#include <boost/test/tools/floating_point_comparison.hpp>
+#endif
+
+#include <dune/common/version.hh>
+#include <dune/grid/common/mcmgmapper.hh>
+
+#include <opm/input/eclipse/Deck/Deck.hpp>
+#include <opm/input/eclipse/EclipseState/Grid/EclipseGrid.hpp>
+#include <opm/input/eclipse/Parser/Parser.hpp>
+#include <opm/grid/CpGrid.hpp>
+#include <opm/grid/LookUpCellCentroid.hh>
+
+#include <sstream>
+#include <iostream>
+
+struct Fixture
+{
+    Fixture()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        Dune::MPIHelper::instance(m_argc, m_argv);
+        Opm::OpmLog::setupSimpleDefaultLogging();
+    }
+
+    static int rank()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        return Dune::MPIHelper::instance(m_argc, m_argv).rank();
+    }
+};
+
+BOOST_GLOBAL_FIXTURE(Fixture);
+
+void createEclGridCpGrid_and_checkCuboidShape(const std::string& deckString)
+{
+    Opm::Parser parser;
+    const auto deck = parser.parseString(deckString);
+
+    Dune::CpGrid grid;
+    Opm::EclipseGrid eclGrid(deck);
+
+    grid.processEclipseFormat(&eclGrid, nullptr, false, false, false);
+
+    const auto& leafGridView = grid.chooseData()[0]; // Leaf Grid coincides with level 0, in this case. 
+    BOOST_CHECK_THROW(leafGridView->checkCuboidShape({0}), std::logic_error);
+}
+
+BOOST_AUTO_TEST_CASE(nonCuboidCell)
+{/*
+   Cell corners:                     COORD
+   0 {0, 0, 0}                       line 1: corners 0 and 4
+   1 {1, 0, 0}                       line 2: corners 1 and 5
+   2 {0, 1, 0}                       line 3: corners 2 and 6
+   3 {1, 1, 0}                       line 4: corners 3 and 7
+   4 {0, 0, 1}
+   5 {1, 0, 0}  coincides with 1
+   6 {0, 1, 1}
+   7 {1, 1, 0}  coincides with 3 */
+    const std::string deckString =
+        R"(RUNSPEC
+        DIMENS
+        1  1  1 /
+        GRID
+        COORD
+        0 0 0  0 0 1
+        1 0 0  1 0 0
+        0 1 0  0 1 1
+        1 1 0  1 1 0
+        /
+        ZCORN
+        4*0
+        4*1
```

Comment:
Looks like a perfect cube to me. Shouldn't one of the top or bottom coordinates be different from 0 or 1?

Reply:
line 108 and line 110 have the same corner twice. That meant to be gluing corner  '1' (from bottom face) with corner '5' (from top face), and  corner  '3' (from bottom face) with corner '7' (from top face).

Then, how should that part of the string be to reflex this shape with 6 different corners?

Thanks for finding the issue!

Reply:
ZCORN is the depth of each point on the lines defined by COORDS.

Reply:
Yes, thanks. I checked the manual. Still need to figure how to pass the 8 corners (with 2 repeated ones) out, so that I get the non-cuboid shape gluing corner '1' with '5' and '3' with '7'